go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -114,44 +102,47 @@ func (npw *nodePortWatcher) AddService(service *kapi.Service) {
 						npw.gwBridge, stderr, err, flow)
 				}
 			}
-		}
-		for _, externalIP := range service.Spec.ExternalIPs {
-			if err := util.ValidatePort(svcPort.Protocol, svcPort.Port); err != nil {
-				klog.Errorf("Skipping service add for svc: %s, err: %v", svcPort.Name, err)
-				continue
-			}
-			flowProtocol := protocol
-			nw_dst := "nw_dst"
-			if utilnet.IsIPv6String(externalIP) {
-				flowProtocol = protocol6
-				nw_dst = "ipv6_dst"
-			}
-			_, stderr, err := util.RunOVSOfctl("add-flow", npw.gwBridge,
-				fmt.Sprintf("priority=100, in_port=%s, %s, %s=%s, tp_dst=%d, actions=%s",
-					npw.ofportPhys, flowProtocol, nw_dst, externalIP, svcPort.Port, npw.ofportPatch))
-			if err != nil {
-				klog.Errorf("Failed to add openflow flow on %s for ExternalIP: "+
-					"%s, stderr: %q, error: %v", npw.gwBridge, externalIP, stderr, err)
+
+			for _, externalIP := range service.Spec.ExternalIPs {
+				if err := util.ValidatePort(svcPort.Protocol, svcPort.Port); err != nil {
+					klog.Errorf("Skipping service add for svc: %s, err: %v", svcPort.Name, err)
+					continue
+				}
+				flowProtocol := protocol
+				nwDst := "nw_dst"
+				if utilnet.IsIPv6String(externalIP) {
+					flowProtocol = protocol6
+					nwDst = "ipv6_dst"
+				}
+				_, stderr, err := util.RunOVSOfctl("add-flow", npw.gwBridge,
+					fmt.Sprintf("priority=100, in_port=%s, %s, %s=%s, tp_dst=%d, actions=%s",
+						npw.ofportPhys, flowProtocol, nwDst, externalIP, svcPort.Port, npw.ofportPatch))
+				if err != nil {
+					klog.Errorf("Failed to add openflow flow on %s for ExternalIP: "+
+						"%s, stderr: %q, error: %v", npw.gwBridge, externalIP, stderr, err)
+				}
 			}
 		}
 	}
 
 	addSharedGatewayIptRules(service, npw.nodeIP)
 }
 
-func (npw *nodePortWatcher) UpdateService(new, old *kapi.Service) {
-	if reflect.DeepEqual(new.Spec, old.Spec) {
+func (npw *nodePortWatcher) UpdateService(old, new *kapi.Service) {
+	if reflect.DeepEqual(new.Spec.Ports, old.Spec.Ports) &&
+		reflect.DeepEqual(new.Spec.ExternalIPs, old.Spec.ExternalIPs) &&
+		reflect.DeepEqual(new.Spec.ClusterIP, old.Spec.ClusterIP) &&
+		reflect.DeepEqual(new.Spec.Type, old.Spec.Type) &&
+		reflect.DeepEqual(new.Status.LoadBalancer.Ingress, old.Status.LoadBalancer.Ingress) {
+		klog.V(5).Infof("Skipping service update for: %s as change does not apply to any of .Spec.Ports, "+
+			".Spec.ExternalIP, .Spec.ClusterIP, .Spec.Type, .Status.LoadBalancer.Ingress", new.Name)
 		return
 	}
 	npw.DeleteService(old)
 	npw.AddService(new)
 }
 
 func (npw *nodePortWatcher) DeleteService(service *kapi.Service) {
-	if !util.ServiceTypeHasNodePort(service) && len(service.Spec.ExternalIPs) == 0 {
```

Comment:
the services that are not filtered by this are a noop ... or it should be :)

Reply:
as we discussed this check isnt adequate because ingress works on service type cluster ip. I'll change it to be the same check you suggested in the previous comment:

```
	// don't process headless service or services that doesn't have NodePorts or ExternalIPs
	if !util.ServiceTypeHasClusterIP(service) || !util.IsClusterIPSet(service) {
		return
	}
```

Reply:
Actually the node informer filters out headless services, so this isn't really needed.

Reply:
heh actually it claims in the comment it filters headless services for service watcher, but it really only filters headless services for endpoint watcher @squeed :

```
	// For Services and Endpoints, pre-populate the shared Informer with one that
	// has a label selector excluding headless services.
	wf.iFactory.InformerFor(&kapi.Service{}, func(c kubernetes.Interface, resyncPeriod time.Duration) cache.SharedIndexInformer {
		return v1coreinformers.NewFilteredServiceInformer(
			c,
			kapi.NamespaceAll,
			resyncPeriod,
			cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
			noAlternateProxySelector())
	})

	wf.iFactory.InformerFor(&kapi.Endpoints{}, func(c kubernetes.Interface, resyncPeriod time.Duration) cache.SharedIndexInformer {
		return v1coreinformers.NewFilteredEndpointsInformer(
			c,
			kapi.NamespaceAll,
			resyncPeriod,
			cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
			noHeadlessServiceSelector())
	})

```